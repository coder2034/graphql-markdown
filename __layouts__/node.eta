<%~ includeFile('./partials/description', it) %>

<% if (['Query','Mutation','Subscription'].includes(it.kind)) { %>
<%~ includeFile('./operation', {...it, operation: it.kind.toLowerCase() } ) %>
<% } else if (it.kind === "ScalarTypeDefinition") { %>
<%~ includeFile('./scalar', it) %>
<% } else { %>
<%~ includeFile('./object', it) %>
<% } %>
