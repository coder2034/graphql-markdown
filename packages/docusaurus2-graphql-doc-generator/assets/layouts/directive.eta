<%~ includeFile('./partials/description', it) %>

```graphql
directive <%~ it.name %> 
```
