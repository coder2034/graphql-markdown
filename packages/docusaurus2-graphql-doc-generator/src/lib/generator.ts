import { promises as fs } from "fs";
import path from "path";

import prettier from "prettier";

import { generateMarkdownFromSchema } from "@edno/graphql-markdown";

import { PluginOptions, Sidebar } from "../types";

const SIDEBAR_FILE = "sidebar-schema.js";

const prettifyJSON = (content: string): string => {
  return prettier.format(content, { parser: "json" });
};

const prettifyMarkdown = (content: string): string => {
  return prettier.format(content, { parser: "markdown" });
};

const renderHomepage = async (
  homepageLocation: string,
  options: PluginOptions
): Promise<string> => {
  const homepage = await fs.readFile(homepageLocation, "utf8");

  const filePath = path.join(options.rootPath, path.basename(homepageLocation));
  await fs.writeFile(filePath, prettifyMarkdown(homepage), "utf8");

  return path.relative("./", filePath);
};

const renderSidebar = async ({
  rootPath,
  baseURL,
}: PluginOptions): Promise<string> => {
  const sidebar: Record<string, Sidebar> = {
    schemaSidebar: [
      {
        dirName: baseURL,
        type: "autogenerated",
      },
    ],
  };

  const filePath = path.join(rootPath, SIDEBAR_FILE);
  await fs.writeFile(filePath, prettifyJSON(JSON.stringify(sidebar)), "utf8");

  return path.relative("./", filePath);
};

export const generateDocFromSchema = async (
  schemaLocation: string,
  options: PluginOptions
): Promise<{ pages: number; sidebar: string }> => {
  const pages = await generateMarkdownFromSchema({ saveToFiles: true });

  // eslint-disable-next-line no-magic-numbers
  if (typeof pages === "undefined" || pages.length === 0) {
    throw new Error("Parsing error, no pages generated");
  }

  if (typeof options === "undefined") {
    throw new Error("Parsing error, no pages generated");
  }

  await renderHomepage(options.homepage, options);
  const sidebarPath = await renderSidebar(options);

  return { pages: pages.length, sidebar: sidebarPath };
};
