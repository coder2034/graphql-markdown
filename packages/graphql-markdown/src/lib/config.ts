import {
  GraphQLExtensionDeclaration,
  GraphQLProjectConfig,
  loadConfig,
} from "graphql-config";
import { DocumentNode } from "graphql";

import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
import { JsonFileLoader } from "@graphql-tools/json-file-loader";
import { UrlLoader } from "@graphql-tools/url-loader";

import { ExtensionAPI, LoadConfigOptions, Maybe } from "..";

const EXTENSION_NAME = "graphql-markdown" as const;

export const OPTION = {
  LAYOUTS: "layouts",
  MDX: "mdx",
  OUTPUT: "output",
};

export type ConfigurationOptions = {
  excludes?: readonly string[];
  layouts?: string;
  mdx?: boolean;
  output?: string;
};

const defaultOptions: ConfigurationOptions = {
  excludes: [] as const,
  [OPTION.LAYOUTS]: "./layouts" as const,
  [OPTION.OUTPUT]: false as const,
  output: "./output" as const,
} as const;

const setFileLoaderExtension: GraphQLExtensionDeclaration = (
  api: ExtensionAPI
) => {
  [new GraphQLFileLoader(), new JsonFileLoader(), new UrlLoader()].forEach(
    (loader) => {
      return api.loaders.schema.register(loader);
    }
  );
  return { name: EXTENSION_NAME };
};

const loadConfiguration = async (
  options?: LoadConfigOptions
): Promise<Maybe<GraphQLProjectConfig>> => {
  const config = await loadConfig({
    ...options,
    extensions: [setFileLoaderExtension],
  });
  return config?.getDefault();
};

export class Configuration {
  static project: Maybe<GraphQLProjectConfig>;
  static extension: ConfigurationOptions;

  static load = async (options?: LoadConfigOptions): Promise<void> => {
    const configuration = await loadConfiguration(options);
    Configuration.project = configuration;
    Configuration.extension = {
      ...defaultOptions,
      ...Configuration.project?.extension(EXTENSION_NAME),
    };
  };

  static get = (name: string): string => {
    return Configuration.extension[name];
  };

  static set = (name: string, value: string): void => {
    Configuration.extension[name] = value;
  };

  // eslint-disable-next-line require-await
  static schema = async (): Promise<Maybe<DocumentNode>> => {
    return Configuration.project!.getSchema("DocumentNode");
  };
}
