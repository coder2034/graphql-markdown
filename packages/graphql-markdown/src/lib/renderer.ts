import * as path from "path";

import * as Eta from "eta";
import * as prettier from "prettier";

import { __basedir, getConfigurationOption } from "../";

export const renderNode = async (node: any): Promise<string> => {
  const layout = path.resolve(
    __basedir,
    getConfigurationOption("layouts"),
    "layout"
  );
  const result = (await Eta.renderFile(layout, node)) as string;

  return prettier.format(result, { parser: "markdown" });
};
