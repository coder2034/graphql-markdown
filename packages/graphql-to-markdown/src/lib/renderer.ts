import * as path from "path";
import * as fs from "fs-extra";
import { pull } from "lodash";

import { toSlug, startCase, hasOwnProperty } from "./utils";
import { prettifyJavascript } from "./prettier";
import { Printer } from "./printer";

const SIDEBAR = "sidebar-schema.js";
const HOMEPAGE_ID = "schema";

export class Renderer {
  constructor(readonly printer: Printer, readonly outputDir: string, readonly baseURL: string) {
    this.emptyOutputDir();
  }

  emptyOutputDir() {
    fs.emptyDirSync(this.outputDir);
  }

  async renderRootTypes(name: string, type: any) {
    let pages: ({ category: string; slug: string } | undefined)[] = [];
    if (type) {
      const slug = toSlug(name);
      const dirPath = path.join(this.outputDir, slug);
      if (Array.isArray(type)) {
        type = type.reduce(function (r, o) {
          if (o && o.name) r[o.name] = o;
          return r;
        }, {});
      }

      fs.ensureDir(dirPath);

      await Promise.all(
        Object.keys(type).map((name: string) =>
          this.renderTypeEntities(dirPath, name, type[name]),
        ),
      ).then((p) => {
        pages = [...p];
      });
    }
    return pages;
  }

  async renderTypeEntities(dirPath: string, name: string, type: any) {
    if (type) {
      const fileName = toSlug(name);
      const filePath = path.join(dirPath, `${fileName}.mdx`);
      const content = this.printer.printType(fileName, type);
      await fs.outputFile(filePath, content, "utf8");
      const page: any = path
        .relative(this.outputDir, filePath)
        .match(
          /(?<category>[A-z][A-z0-9-]*)\/(?<pageId>[A-z][A-z0-9-]*).mdx?$/,
        );
      const slug = path.join(page.groups.category, page.groups.pageId);
      return { category: startCase(page.groups.category), slug: slug };
    }
  }

  async renderSidebar(pages: any[]) {
    const filePath = path.join(this.outputDir, SIDEBAR);
    const content = prettifyJavascript(`module.exports = {
          schemaSidebar:
          ${JSON.stringify(this.generateSidebar(pages))}
        };`);
    fs.outputFileSync(filePath, content, "utf8");
    return path.relative("./", filePath);
  }

  generateSidebar(pages: any[]) {
    let graphqlSidebar = [
      { type: "doc", id: path.join(this.baseURL, HOMEPAGE_ID) },
    ];
    pages.map((page: { category: any; slug: any }) => {
      const category: any = graphqlSidebar.find(
        (entry: any) => "label" in entry && page.category == entry.label,
      );
      const items = hasOwnProperty(category, "items") ? category?.items : [];
      const slug = path.join(this.baseURL, page.slug);
      graphqlSidebar = [
        ...pull(graphqlSidebar, category),
        {
          type: "category",
          label: page.category,
          items: [...items, slug].sort(),
        },
      ];
    });
    return graphqlSidebar;
  }
}
