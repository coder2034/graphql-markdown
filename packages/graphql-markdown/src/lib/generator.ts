import { Configuration, ConfigurationOptions } from "./config";
import { LoadConfigOptions, ParsedNode, Result } from "..";
import { renderNode, saveMarkdownFile } from "./render";
import { parseSchema } from "./parser";

const loadNodes = async (
  options?: LoadConfigOptions
): Promise<ParsedNode[]> => {
  await Configuration.load(options);

  const schema = await Configuration.schema();

  if (typeof schema === "undefined") {
    throw new Error("An error occurred while loading the schema");
  }

  return parseSchema(schema);
};

type Options = {
  saveToFiles?: boolean;
  loaderOptions?: LoadConfigOptions;
  generatorOptions?: ConfigurationOptions;
};

export const generateMarkdownFromSchema = async (
  options?: Options
): Promise<Result[]> => {
  const nodes = await loadNodes(options?.loaderOptions);

  const processor =
    options?.saveToFiles === true ? saveMarkdownFile : renderNode;

  const results = await Promise.all(
    nodes.map(async (node: ParsedNode): Promise<Result> => {
      const result = await processor(node, options?.generatorOptions);
      return {
        name: node.name,
        type: node.type,
        /* eslint-disable prettier/prettier */
        ...(options?.saveToFiles === true
          ? { filepath: result }
          : { markdown: result }),
        /* eslint-enable */
      };
    })
  );

  return results;
};
