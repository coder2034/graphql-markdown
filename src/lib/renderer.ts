import * as path from "path";

import * as Eta from "eta";
import * as prettier from "prettier";

import { getConfigurationOption } from "./config";

export const renderNode = async (node: any): Promise<string> => {
  const layout = path.resolve(
    process.cwd(),
    getConfigurationOption("layouts"),
    "layout"
  );
  const result = (await Eta.renderFile(layout, node)) as string;

  return prettier.format(result, { parser: "markdown" });
};
