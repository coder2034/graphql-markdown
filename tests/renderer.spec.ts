import { renderNode } from "../src/lib/renderer";

import kindMutation from "./__data__/node/mutation";
import kindSubscription from "./__data__/node/subscription";
import kindObject from "./__data__/node/object";
import kindInput from "./__data__/node/input";
import kindScalar from "./__data__/node/scalar";
import kindQuery from "./__data__/node/query";
import { directiveDeprecated, directiveSpecifiedBy } from "./__data__/node/directive";

describe("renderNode", () => {
  describe("Query", () => {
    it("should render Query", async () => {
      expect.hasAssertions();

      const result = await renderNode(kindQuery);
      expect(result).toMatchSnapshot();
});
  });

  describe("Mutation", () => {
    it("should render Mutation", async () => {
      expect.hasAssertions();

      const result = await renderNode(kindMutation);
      expect(result).toMatchSnapshot();
    });
  });

  describe("Subscription", () => {
    it("should render Subscription", async () => {
      expect.hasAssertions();

      const result = await renderNode(kindSubscription);
      expect(result).toMatchSnapshot();
    });
  });

  describe("Enum", () => {
    it("should render Enum", () => {});
  });

  describe("Object", () => {
    it("should render Object", async () => {
      expect.hasAssertions();

      const result = await renderNode(kindObject);
      expect(result).toMatchSnapshot();
    });
  });

  describe("Input", () => {
    it("should render Input", async () => {
      expect.hasAssertions();

      const result = await renderNode(kindInput);
      expect(result).toMatchSnapshot();
    });
  });

  describe("Scalar", () => {
    it("should render Scalar without description", async () => {
      expect.hasAssertions();

      const result = await renderNode({
        ...kindScalar,
        description: undefined
      });
      expect(result).toMatchSnapshot();
    });

    it("should render Scalar with description supporting Markdown notation", async () => {
      expect.hasAssertions();

      const result = await renderNode(kindScalar);
      expect(result).toMatchSnapshot();
    });

    it("should render Scalar with specifiedBy", async () => {
      expect.hasAssertions();

      const result = await renderNode({...kindScalar, directives: [directiveSpecifiedBy]});
      expect(result).toMatchSnapshot();
    });
  });

  describe("Interface", () => {
    it("should render Interface", () => {});
  });

  describe("Union", () => {
    it("should render Union", () => {});
  });
});
