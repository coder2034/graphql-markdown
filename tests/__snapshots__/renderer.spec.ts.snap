// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderNode Directive should render Directive 1`] = `
"## DirectiveDefinition \`@auth\`

\`\`\`graphql
directive @auth
\`\`\`
"
`;

exports[`renderNode Enum should render Enum 1`] = `
"## EnumTypeDefinition \`MediaType\`

\`\`\`graphql
EnumTypeDefinition MediaType
\`\`\`
"
`;

exports[`renderNode Input should render Input 1`] = `
"## InputObjectTypeDefinition \`TweetContent\`

\`\`\`graphql
InputObjectTypeDefinition TweetContent
\`\`\`

### Fields

#### content (\`String\`)

#### body (\`String\`)
"
`;

exports[`renderNode Interface should render Interface 1`] = `
"## InterfaceTypeDefinition \`Node\`

\`\`\`graphql
InterfaceTypeDefinition Node
\`\`\`

### Fields

#### id (\`ID\`)
"
`;

exports[`renderNode Mutation should render Mutation 1`] = `
"## Mutation \`markTweetRead\`

\`\`\`graphql
mutation markTweetRead(
  id: ID!
): Boolean!
\`\`\`

### Arguments

#### id (\`ID\`)

### Type

#### [\`Boolean\`]
"
`;

exports[`renderNode Mutation should render Mutation with directive deprecated 1`] = `
"## Mutation \`markTweetRead\`

\`\`\`graphql
mutation markTweetRead(
  id: ID!
): Boolean!
\`\`\`

### Arguments

#### id (\`ID\`)

### Type

#### [\`Boolean\`]
"
`;

exports[`renderNode Object should render Object 1`] = `
"## ObjectTypeDefinition \`Media\`

\`\`\`graphql
ObjectTypeDefinition Media
\`\`\`

### Fields

#### url (\`Url\`)

#### sri (\`SRI\`)
"
`;

exports[`renderNode Query should render Query 1`] = `
"## Query \`Tweet\`

\`\`\`graphql
query Tweet(
  id: ID!
): Tweet
\`\`\`

### Arguments

#### id (\`ID\`)

### Type

#### [\`Tweet\`]
"
`;

exports[`renderNode Query should render Query with directive deprecated 1`] = `
"## Query \`Tweet\`

\`\`\`graphql
query Tweet(
  id: ID!
): Tweet
\`\`\`

### Arguments

#### id (\`ID\`)

### Type

#### [\`Tweet\`]
"
`;

exports[`renderNode Scalar should render Scalar with description supporting Markdown notation 1`] = `
"## ScalarTypeDefinition \`SRI\`

An \`integrity\` value beginning with at least one string, with each string including a prefix indicating a particular hash algorithm (e.g. \`sha256\`, \`sha384\`, \`sha512\`), followed by a dash \`-\`, and ending with the actual base64-encoded hash.

\`\`\`graphql
scalar SRI
\`\`\`
"
`;

exports[`renderNode Scalar should render Scalar with directive specifiedBy 1`] = `
"## ScalarTypeDefinition \`SRI\`

An \`integrity\` value beginning with at least one string, with each string including a prefix indicating a particular hash algorithm (e.g. \`sha256\`, \`sha384\`, \`sha512\`), followed by a dash \`-\`, and ending with the actual base64-encoded hash.

\`\`\`graphql
scalar SRI
\`\`\`
"
`;

exports[`renderNode Scalar should render Scalar without description 1`] = `
"## ScalarTypeDefinition \`SRI\`

\`\`\`graphql
scalar SRI
\`\`\`
"
`;

exports[`renderNode Subscription should render Subscription 1`] = `
"## Subscription \`Notifications\`

\`\`\`graphql
subscription Notifications(
  limit: Int!
): Notification[]!
\`\`\`

### Arguments

#### limit (\`Int\`)

### Type

#### [\`Notification\`]
"
`;

exports[`renderNode Subscription should render Subscription with directive deprecated 1`] = `
"## Subscription \`Notifications\`

\`\`\`graphql
subscription Notifications(
  limit: Int!
): Notification[]!
\`\`\`

### Arguments

#### limit (\`Int\`)

### Type

#### [\`Notification\`]
"
`;

exports[`renderNode Union should render Union 1`] = `
"## UnionTypeDefinition \`TweetResponse\`

\`\`\`graphql
UnionTypeDefinition TweetResponse
\`\`\`
"
`;
